Source/Core/GoogleEarthEnterpriseTerrainProvider.js

```diff
@@ -0,0 +1,587 @@
+     * @param {Boolean} [throttleRequests=true] True if the number of simultaneous requests should be limited,
+     *                  or false if the request should be initiated regardless of the number of requests
+     *                  already in progress.
+     * @returns {Promise.<TerrainData>|undefined} A promise for the requested geometry.  If this method
+     *          returns undefined instead of a promise, it is an indication that too many requests are already
+     *          pending and the request will be retried later.
+     *
+     * @exception {DeveloperError} This function must not be called before {@link GoogleEarthEnterpriseProvider#ready}
+     *            returns true.
+     */
+    GoogleEarthEnterpriseTerrainProvider.prototype.requestTileGeometry = function(x, y, level, throttleRequests) {
+        //>>includeStart('debug', pragmas.debug)
+        if (!this._ready) {
+            throw new DeveloperError('requestTileGeometry must not be called before the terrain provider is ready.');
+        }
+        //>>includeEnd('debug');
+
+        var quadKey = GoogleEarthEnterpriseMetadata.tileXYToQuadKey(x, y, level);
+        var terrainCache = this._terrainCache;
+        var metadata = this._metadata;
+        var info = metadata.getTileInformationFromQuadKey(quadKey);
+
+        // Check if this tile is even possibly available
+        if (!defined(info)) {
+            return when.reject(new RuntimeError('Terrain tile doesn\'t exist'));
+        }
+
+        var terrainState = info.terrainState;
+        if (!defined(terrainState)) {
+            // First time we have tried to load this tile, so set terrain state to UNKNOWN
+            terrainState = info.terrainState = TerrainState.UNKNOWN;
+        }
+
+        // If its in the cache, return it
+        var buffer = terrainCache.get(quadKey);
+        if (defined(buffer)) {
+            return new GoogleEarthEnterpriseTerrainData({
+                buffer : buffer,
+                childTileMask : computeChildMask(quadKey, info, metadata)
+            });
+        }
+
+        // Clean up the cache
+        terrainCache.tidy();
+
+        // We have a tile, check to see if no ancestors have terrain or that we know for sure it doesn't
+        if (!info.ancestorHasTerrain) {
+            // We haven't reached a level with terrain, so return the ellipsoid
+            return new HeightmapTerrainData({
+                buffer : new Uint8Array(16 * 16),
+                width : 16,
+                height : 16
+            });
+        } else if (terrainState === TerrainState.NONE) {
+            // Already have info and there isn't any terrain here
+            return when.reject(new RuntimeError('Terrain tile doesn\'t exist'));
+        }
+
+        // Figure out where we are getting the terrain and what version
+        var parentInfo;
+        var q = quadKey;
+        var terrainVersion = -1;
+        if (terrainState !== TerrainState.NONE) {
```

Comment:
I don't think this `if` is necessary.  If `terrainState` were `TerrainState.NONE`, we would have returned above.  And even if not, entering the switch won't do any harm.